### ISSUE TITLE
Console says No SLF4J providers were found.

### ISSUE THREAD

**Issue by Angular-Angel** [MinecraftForge/MinecraftForge#7904]
**Minecraft Version:** 1.17.1

**Forge Version:** 37.0.1

**Logs:** {Link(s) to GitHub Gist with full latest.log and/or crash report}

**Steps to Reproduce:**
 1. Install a fresh MDK project
 2. Run the project
 3. Check the console output.

**Description of issue:**
When I run my project and try and use the logger, I don't get anything written in the console, but when I go and read through the console, I find the following: 

```
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.

```

**Comment by Angular-Angel** [comment]
Nvm, log does write to console. My mistake!

**Comment by sciwhiz12** [comment]
This seems to be an issue with how [ModLauncher][ml]'s [`TransformingClassLoader`][ml-tcl] works with resources, which is managed by its superclass, [securejarhandler][sjh]'s [`ModuleClassLoader`][sjh-mcl]. 

### Some background

When a resource is queried which does not belong in a specific module -- either because no module claims ownership of the package it belongs to or it has no package -- it will serach for the resource in each module registered in the class loader (the `resolvedRoots` map of module names to module references). However, it _does not_ search for the resource in the fallback classloader, so resources from that are inaccessible to the classes in the module class loader.

In Minecraft Forge, there is a hierarchy of 4 module layers where different code resides in, as represented in the [`IModuleLayerManager.Layer` enum][ml-layer-enum]. In this situation, we are concerned with two module layers: the boot layer (`BOOT`) and the game layer (`GAME`). The game layer's classloader (which is a `TransformingClassLoader`) has a fallback classloader of the boot layer's classloader (which is a `ModuleClassLoader`, the one set by [bootstraplauncher][bsl]). 

### The problem

When the game calls upon `ResourceLeakDetector`, which leads to a chain where Netty loads the SLF4J `LoggerFactory`. The `LoggerFactory` then uses `ServiceLoader` to find and load services of `org.slf4j.ILoggerFactory` through the current thread's context class loader.

The problem here is that, when this happens, the application is running on the game layer, and so the context class loader for the current thread is the _game_ layer's classloader. Meanwhile, the service definition of `org.apache.logging.slf4j.Log4jLoggerFactory` (which is stored as a resource under `META-INF/services` of `log4j-slf4j18-impl-2.15.0.jar`) is accessible through the _boot_  layer's classloader. This means that when the `ServiceLoader` queries the game layer's classloader for service definition resources of `META-INF/services/org.slf4j.ILoggerFactory`, it cannot find them since they are on the boot layer's classloader. This then causes SLF4J to log the given message and default to the no-op logger factory.

### Possible solutions

There are two possible solutions for this issue, then:

1. Manually set the logging factory used by Netty to the Log4J implementation, rather than the SLF4J implementation (which it defaults to). This can be done by inserting the following call before the first reference to Netty: `InternalLoggerFactory.setDefaultFactory(Log4JLoggerFactory.INSTANCE);`. However, this does not fix the root issue of the resources, and this issue will still occur if something else later in the application calls upon SLF4J.

1. Create a implementation of an `ILoggerFactory` service, whose definition is accessible by the game layer's classloader, which redirects to `org.apache.logging.slf4j.SLF4JServiceProvider` (either by referencing directly, using conditional classloading, or creating a `ServiceLoader` with the correct module layer to resolve a proper `ILoggerFactory` service. This does not fix the root issue of the resources, but will fix this issue even if something else later calls upon SLF4J.

1. Amend the design of `ModuleClassLoader` such that, when querying for resources, it includes the resources from the fallback classloader. This would necessitate a change in securejarhandler, which may take an extended amount of time to integrate in but will resolve the issue for good.[^1]

Of the three above, I prefer the second choice (implementation of a redirecting `ILoggerFactory`) until @cpw or another of the core team have the time to review and implement the suggestion of amending `ModuleClassLoader` as mentioned above.

[ml]: https://github.com/cpw/modlauncher
[ml-tcl]: https://github.com/cpw/modlauncher/blob/8569cdfcbc41cf97d7fcf94e34789072af64c0a8/src/main/java/cpw/mods/modlauncher/TransformingClassLoader.java
[sjh]: https://github.com/MinecraftForge/securejarhandler
[sjh-mcl]: https://github.com/MinecraftForge/securejarhandler/blob/d356796b1d64bb716850f649e8ea76a1cea456ec/src/main/java/cpw/mods/cl/ModuleClassLoader.java
[sjh-mcl-loadClass]: https://github.com/MinecraftForge/securejarhandler/blob/d356796b1d64bb716850f649e8ea76a1cea456ec/src/main/java/cpw/mods/cl/ModuleClassLoader.java#L127-L134
[ml-layer-enum]: https://github.com/cpw/modlauncher/blob/2880c33583207895f7d9a1377c46ffd8c51867e2/src/main/java/cpw/mods/modlauncher/api/IModuleLayerManager.java#L26-L41
[bsl]: https://github.com/MinecraftForge/bootstraplauncher

[^1]: This assumes that it is not intended behavior for the module class loader to not include resources from the fallback classloader.

**Comment by rsmyst** [comment]
Pls find a solution quickly, i'm unable to play on forge optifine in 1.17.1 version and the latest forge 37.1.1 due to this error

**Comment by sciwhiz12** [comment]
@rsmyst:
1. This issue does not affect the playability of any version of Forge.
2. The issue you are facing may be caused by OptiFine itself. Please try removing OptiFine first.
3. Minecraft 1.17.1 is nearing the end of its support lifecycle (pending a grace period of give or take a month). Please update to the latest version (or the LTS version) to receive support.
4. This issue tracker is for feature requests and final bug reports. If you are having difficulties, please visit the [Forge Support Forum](https://forums.minecraftforge.net/forum/18-support-bug-reports/) or the official Forge discord to ask for support.

**Comment by ExE-Boss** [comment]
This is making it difficult to port **[NOVA](https://github.com/NOVA-Team/NOVA-Monorepo)** to **Minecraft 1.18.1**, as it uses **SLF4J** extensively.

**Comment by sciwhiz12** [comment]
Fixed in 1.18.2 by [`SLF4JFixerLaunchPluginService`](https://github.com/MinecraftForge/MinecraftForge/blob/8ade7bbaebdaa263b0391fd165d5e4e033b52c60/fmlloader/src/main/java/net/minecraftforge/fml/loading/log4j/SLF4JFixerLaunchPluginService.java) which sets the correct context class loader and triggers SLF4J to bind to a logging factory immediately.

**Comment by bogadombc** [comment]
I solved this follow the steps :  my project : Maven JSF Primefaces
- open eclipse marketplace
- open Installed
- check if there is an update for  " Eclipse m2e - Maven support in Eclipse IDE Latest "
